File: apps/desktop/src-tauri/src/commands/tag.rs
use tauri::State;

use crate::{
    errors::{CommandResult, IntoCommandResult},
    models::tag::{
        DeleteTagGroupPayload, DeleteTagPayload, SaveTagGroupPayload,
        SaveTagPayload, TagIndex,
    },
    services::TagService,
};

#[tauri::command]
pub async fn load_tag_index(
    tag_service: State<'_, TagService>,
) -> CommandResult<TagIndex> {
    tag_service.load_tag_index().await.into_command()
}

#[tauri::command]
pub async fn save_tag_group(
    payload: SaveTagGroupPayload,
    tag_service: State<'_, TagService>,
) -> CommandResult<TagIndex> {
    tag_service.save_tag_group(payload).await.into_command()
}

#[tauri::command]
pub async fn delete_tag_group(
    payload: DeleteTagGroupPayload,
    tag_service: State<'_, TagService>,
) -> CommandResult<TagIndex> {
    tag_service.delete_tag_group(payload).await.into_command()
}

#[tauri::command]
pub async fn save_tag(
    payload: SaveTagPayload,
    tag_service: State<'_, TagService>,
) -> CommandResult<TagIndex> {
    tag_service.save_tag(payload).await.into_command()
}

#[tauri::command]
pub async fn delete_tag(
    payload: DeleteTagPayload,
    tag_service: State<'_, TagService>,
) -> CommandResult<TagIndex> {
    tag_service.delete_tag(payload).await.into_command()
}

File: apps/desktop/src/shared/lib/ipc/tag.ts
import type {
  DeleteTagGroupPayload,
  DeleteTagPayload,
  SaveTagGroupPayload,
  SaveTagPayload,
  TagIndex,
} from '../../types';
import { invokeCamel } from './core';

export const tagIpc = {
  loadIndex: (): Promise<TagIndex> => invokeCamel('load_tag_index'),
  saveGroup: (payload: SaveTagGroupPayload) => invokeCamel('save_tag_group', { payload }),
  deleteGroup: (payload: DeleteTagGroupPayload) => invokeCamel('delete_tag_group', { payload }),
  saveTag: (payload: SaveTagPayload) => invokeCamel('save_tag', { payload }),
  deleteTag: (payload: DeleteTagPayload) => invokeCamel('delete_tag', { payload }),
};
